src/main/flight/pid.c

```diff
@@ -263,74 +337,40 @@ void updatePIDCoefficients(const pidProfile_t *pidProfile, const controlRateConf
         return;
     }
 
-    // Calculate TPA factor - different logic for airplanes and multirotors
-    if (STATE(FIXED_WING)) {
-        // tpa_rate is amount of curve TPA applied to PIDs
-        // tpa_breakpoint for fixed wing is cruise throttle value (value at which PIDs were tuned)
-        if (controlRateConfig->dynThrPID != 0 && controlRateConfig->tpa_breakpoint > motorConfig->minthrottle) {
-            if (rcCommand[THROTTLE] > motorConfig->minthrottle) {
-                // Calculate TPA according to throttle
-                tpaFactor = 0.5f + ((float)(controlRateConfig->tpa_breakpoint - motorConfig->minthrottle) / (rcCommand[THROTTLE] - motorConfig->minthrottle) / 2.0f);
-
-                // Limit to [0.5; 2] range
-                tpaFactor = constrainf(tpaFactor, 0.5f, 2.0f);
-            }
-            else {
-                tpaFactor = 2.0f;
-            }
-
-            // Attenuate TPA curve according to configured amount
-            tpaFactor = 1.0f + (tpaFactor - 1.0f) * (controlRateConfig->dynThrPID / 100.0f);
-        }
-        else {
-            tpaFactor = 1.0f;
-        }
-    }
-    else {
-        // TPA should be updated only when TPA is actually set
-        if (controlRateConfig->dynThrPID == 0 || rcCommand[THROTTLE] < controlRateConfig->tpa_breakpoint) {
-            tpaFactor = 1.0f;
-        } else if (rcCommand[THROTTLE] < motorConfig->maxthrottle) {
-            tpaFactor = (100 - (uint16_t)controlRateConfig->dynThrPID * (rcCommand[THROTTLE] - controlRateConfig->tpa_breakpoint) / (motorConfig->maxthrottle - controlRateConfig->tpa_breakpoint)) / 100.0f;
-        } else {
-            tpaFactor = (100 - controlRateConfig->dynThrPID) / 100.0f;
-        }
-    }
+    const float tpaFactor = STATE(FIXED_WING) ? calculateFixedWingTPAFactor() : calculateMultirotorTPAFactor();
 
     // PID coefficients can be update only with THROTTLE and TPA or inflight PID adjustments
     //TODO: Next step would be to update those only at THROTTLE or inflight adjustments change
     for (int axis = 0; axis < 3; axis++) {
-        pidState[axis].kP = pidProfile->P8[axis] / FP_PID_RATE_P_MULTIPLIER;
-        pidState[axis].kI = pidProfile->I8[axis] / FP_PID_RATE_I_MULTIPLIER;
-        pidState[axis].kD = pidProfile->D8[axis] / FP_PID_RATE_D_MULTIPLIER;
-
-        // Apply TPA
         if (STATE(FIXED_WING)) {
             // Airplanes - scale all PIDs according to TPA
-            pidState[axis].kP *= tpaFactor;
-            pidState[axis].kI *= tpaFactor;
-            pidState[axis].kD *= tpaFactor * tpaFactor;     // acceleration scales with speed^2
+            pidState[axis].kP  = pidBank()->pid[axis].P / FP_PID_RATE_P_MULTIPLIER  * tpaFactor * tpaFactor;
```

Comment:
This logic is borrowed from APM and PX4 - they both apply scaling squared to error-driven P-term and plain scaling to feed-forward. We'll see if this logic should be changed - we don't have much feedback of PIFF yet.